<a href="https://colab.research.google.com/github/Ella244-18/phase-1-project/blob/master/student.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Final Project Submission

Please fill out:
* Student name:
* Student pace: self paced / part time / full time
* Scheduled project review date/time:
* Instructor name:
* Blog post URL:


In [ ]:
# Your code here - remember to use markdown cells for comments as well!

In [4]:
#Import the necessary libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import warnings

##Loading the data separately.

###-df1 representing the bom.movie_gross data file and df2 represents the title.basics.csv data file
### -df shows the top 5 and bottom 5 rows while df.head shows the top rows and df.tail
### shows the bottom rows of the data

In [16]:
# Let's preview the data

In [15]:
df1 = pd.read_csv('/content/bom.movie_gross.csv')
df

tconst                                primary_title  \
0       tt0063540                                    Sunghursh   
1       tt0066787              One Day Before the Rainy Season   
2       tt0069049                   The Other Side of the Wind   
3       tt0069204                              Sabse Bada Sukh   
4       tt0100275                     The Wandering Soap Opera   
...           ...                                          ...   
146139  tt9916538                          Kuambil Lagi Hatiku   
146140  tt9916622  Rodolpho Teóphilo - O Legado de um Pioneiro   
146141  tt9916706                              Dankyavar Danka   
146142  tt9916730                                       6 Gunn   
146143  tt9916754               Chico Albuquerque - Revelações   

                                     original_title  start_year  \
0                                         Sunghursh        2013   
1                                   Ashad Ka Ek Din        2019   
2                        The Other Side of the Wind        2018   
3                                   Sabse Bada Sukh        2018   
4                             La Telenovela Errante        2017   
...                                             ...         ...   
146139                          Kuambil Lagi Hatiku        2019   
146140  Rodolpho Teóphilo - O Legado de um Pioneiro        2015   
146141                              Dankyavar Danka        2013   
146142                                       6 Gunn        2017   
146143               Chico Albuquerque - Revelações        2013   

        runtime_minutes                genres  
0                 175.0    Action,Crime,Drama  
1                 114.0       Biography,Drama  
2                 122.0                 Drama  
3                   NaN          Comedy,Drama  
4                  80.0  Comedy,Drama,Fantasy  
...                 ...                   ...  
146139            123.0                 Drama  
146140              NaN           Documentary  
146141              NaN                Comedy  
146142            116.0                   NaN  
146143              NaN           Documentary  

[146144 rows x 6 columns]

In [9]:
df2 = pd.read_csv('/content/title.basics.csv')
df

tconst                                primary_title  \
0       tt0063540                                    Sunghursh   
1       tt0066787              One Day Before the Rainy Season   
2       tt0069049                   The Other Side of the Wind   
3       tt0069204                              Sabse Bada Sukh   
4       tt0100275                     The Wandering Soap Opera   
...           ...                                          ...   
146139  tt9916538                          Kuambil Lagi Hatiku   
146140  tt9916622  Rodolpho Teóphilo - O Legado de um Pioneiro   
146141  tt9916706                              Dankyavar Danka   
146142  tt9916730                                       6 Gunn   
146143  tt9916754               Chico Albuquerque - Revelações   

                                     original_title  start_year  \
0                                         Sunghursh        2013   
1                                   Ashad Ka Ek Din        2019   
2                        The Other Side of the Wind        2018   
3                                   Sabse Bada Sukh        2018   
4                             La Telenovela Errante        2017   
...                                             ...         ...   
146139                          Kuambil Lagi Hatiku        2019   
146140  Rodolpho Teóphilo - O Legado de um Pioneiro        2015   
146141                              Dankyavar Danka        2013   
146142                                       6 Gunn        2017   
146143               Chico Albuquerque - Revelações        2013   

        runtime_minutes                genres  
0                 175.0    Action,Crime,Drama  
1                 114.0       Biography,Drama  
2                 122.0                 Drama  
3                   NaN          Comedy,Drama  
4                  80.0  Comedy,Drama,Fantasy  
...                 ...                   ...  
146139            123.0                 Drama  
146140              NaN           Documentary  
146141              NaN                Comedy  
146142            116.0                   NaN  
146143              NaN           Documentary  

[146144 rows x 6 columns]

In [10]:
df1.head(8)

title studio  domestic_gross  \
0                                  Toy Story 3     BV     415000000.0   
1                   Alice in Wonderland (2010)     BV     334200000.0   
2  Harry Potter and the Deathly Hallows Part 1     WB     296000000.0   
3                                    Inception     WB     292600000.0   
4                          Shrek Forever After   P/DW     238700000.0   
5                   The Twilight Saga: Eclipse   Sum.     300500000.0   
6                                   Iron Man 2   Par.     312400000.0   
7                                      Tangled     BV     200800000.0   

  foreign_gross  year  
0     652000000  2010  
1     691300000  2010  
2     664300000  2010  
3     535700000  2010  
4     513900000  2010  
5     398000000  2010  
6     311500000  2010  
7     391000000  2010

In [11]:
df1.tail()

title      studio  domestic_gross foreign_gross  \
3382                    The Quake       Magn.          6200.0           NaN   
3383  Edward II (2018 re-release)          FM          4800.0           NaN   
3384                     El Pacto        Sony          2500.0           NaN   
3385                     The Swan  Synergetic          2400.0           NaN   
3386            An Actor Prepares       Grav.          1700.0           NaN   

      year  
3382  2018  
3383  2018  
3384  2018  
3385  2018  
3386  2018

In [12]:
df2.head(6)

tconst                    primary_title              original_title  \
0  tt0063540                        Sunghursh                   Sunghursh   
1  tt0066787  One Day Before the Rainy Season             Ashad Ka Ek Din   
2  tt0069049       The Other Side of the Wind  The Other Side of the Wind   
3  tt0069204                  Sabse Bada Sukh             Sabse Bada Sukh   
4  tt0100275         The Wandering Soap Opera       La Telenovela Errante   
5  tt0111414                      A Thin Life                 A Thin Life   

   start_year  runtime_minutes                genres  
0        2013            175.0    Action,Crime,Drama  
1        2019            114.0       Biography,Drama  
2        2018            122.0                 Drama  
3        2018              NaN          Comedy,Drama  
4        2017             80.0  Comedy,Drama,Fantasy  
5        2018             75.0                Comedy

In [13]:
df2.tail(4)

tconst                                primary_title  \
146140  tt9916622  Rodolpho Teóphilo - O Legado de um Pioneiro   
146141  tt9916706                              Dankyavar Danka   
146142  tt9916730                                       6 Gunn   
146143  tt9916754               Chico Albuquerque - Revelações   

                                     original_title  start_year  \
146140  Rodolpho Teóphilo - O Legado de um Pioneiro        2015   
146141                              Dankyavar Danka        2013   
146142                                       6 Gunn        2017   
146143               Chico Albuquerque - Revelações        2013   

        runtime_minutes       genres  
146140              NaN  Documentary  
146141              NaN       Comedy  
146142            116.0          NaN  
146143              NaN  Documentary

In [19]:
# Checking the info of the data
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3387 entries, 0 to 3386
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           3387 non-null   object 
 1   studio          3382 non-null   object 
 2   domestic_gross  3359 non-null   float64
 3   foreign_gross   2037 non-null   object 
 4   year            3387 non-null   int64  
dtypes: float64(1), int64(1), object(3)
memory usage: 132.4+ KB


From the above code we see that df1 has **3387** **rows** and **5** **columns**.

We can also see that some columns have missing values.

Studio has **5** missing values, domestic_gross has  **28** and foreign_gross has **1350** missing values


In [20]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 146144 entries, 0 to 146143
Data columns (total 6 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   tconst           146144 non-null  object 
 1   primary_title    146143 non-null  object 
 2   original_title   146122 non-null  object 
 3   start_year       146144 non-null  int64  
 4   runtime_minutes  114405 non-null  float64
 5   genres           140736 non-null  object 
dtypes: float64(1), int64(1), object(4)
memory usage: 6.7+ MB


in df2 we have 146144 rows and 6 columns

In [27]:
# Getting our columns

In [28]:
df1.columns

Index(['title', 'studio', 'domestic_gross', 'foreign_gross', 'year'], dtype='object')

In [29]:
df2.columns

Index(['tconst', 'primary_title', 'original_title', 'start_year',
       'runtime_minutes', 'genres'],
      dtype='object')

In [30]:
# Describing the data

In [32]:
df1.describe()

domestic_gross         year
count    3.359000e+03  3387.000000
mean     2.874585e+07  2013.958075
std      6.698250e+07     2.478141
min      1.000000e+02  2010.000000
25%      1.200000e+05  2012.000000
50%      1.400000e+06  2014.000000
75%      2.790000e+07  2016.000000
max      9.367000e+08  2018.000000

In [33]:
df2.describe()

start_year  runtime_minutes
count  146144.000000    114405.000000
mean     2014.621798        86.187247
std         2.733583       166.360590
min      2010.000000         1.000000
25%      2012.000000        70.000000
50%      2015.000000        87.000000
75%      2017.000000        99.000000
max      2115.000000     51420.000000

###Merging the two Dataframes

In [40]:
# Merging by stacking all rows from both datasets
combined_df = pd.concat([df1,df2], ignore_index= True)
combined_df

title studio  domestic_gross  \
0                                       Toy Story 3     BV     415000000.0   
1                        Alice in Wonderland (2010)     BV     334200000.0   
2       Harry Potter and the Deathly Hallows Part 1     WB     296000000.0   
3                                         Inception     WB     292600000.0   
4                               Shrek Forever After   P/DW     238700000.0   
...                                             ...    ...             ...   
149526                                          NaN    NaN             NaN   
149527                                          NaN    NaN             NaN   
149528                                          NaN    NaN             NaN   
149529                                          NaN    NaN             NaN   
149530                                          NaN    NaN             NaN   

       foreign_gross    year     tconst  \
0          652000000  2010.0        NaN   
1          691300000  2010.0        NaN   
2          664300000  2010.0        NaN   
3          535700000  2010.0        NaN   
4          513900000  2010.0        NaN   
...              ...     ...        ...   
149526           NaN     NaN  tt9916538   
149527           NaN     NaN  tt9916622   
149528           NaN     NaN  tt9916706   
149529           NaN     NaN  tt9916730   
149530           NaN     NaN  tt9916754   

                                      primary_title  \
0                                               NaN   
1                                               NaN   
2                                               NaN   
3                                               NaN   
4                                               NaN   
...                                             ...   
149526                          Kuambil Lagi Hatiku   
149527  Rodolpho Teóphilo - O Legado de um Pioneiro   
149528                              Dankyavar Danka   
149529                                       6 Gunn   
149530               Chico Albuquerque - Revelações   

                                     original_title  start_year  \
0                                               NaN         NaN   
1                                               NaN         NaN   
2                                               NaN         NaN   
3                                               NaN         NaN   
4                                               NaN         NaN   
...                                             ...         ...   
149526                          Kuambil Lagi Hatiku      2019.0   
149527  Rodolpho Teóphilo - O Legado de um Pioneiro      2015.0   
149528                              Dankyavar Danka      2013.0   
149529                                       6 Gunn      2017.0   
149530               Chico Albuquerque - Revelações      2013.0   

        runtime_minutes       genres  
0                   NaN          NaN  
1                   NaN          NaN  
2                   NaN          NaN  
3                   NaN          NaN  
4                   NaN          NaN  
...                 ...          ...  
149526            123.0        Drama  
149527              NaN  Documentary  
149528              NaN       Comedy  
149529            116.0          NaN  
149530              NaN  Documentary  

[149531 rows x 11 columns]

In [41]:
combined_df.shape

(149531, 11)

In [42]:
combined_df.head()

title studio  domestic_gross  \
0                                  Toy Story 3     BV     415000000.0   
1                   Alice in Wonderland (2010)     BV     334200000.0   
2  Harry Potter and the Deathly Hallows Part 1     WB     296000000.0   
3                                    Inception     WB     292600000.0   
4                          Shrek Forever After   P/DW     238700000.0   

  foreign_gross    year tconst primary_title original_title  start_year  \
0     652000000  2010.0    NaN           NaN            NaN         NaN   
1     691300000  2010.0    NaN           NaN            NaN         NaN   
2     664300000  2010.0    NaN           NaN            NaN         NaN   
3     535700000  2010.0    NaN           NaN            NaN         NaN   
4     513900000  2010.0    NaN           NaN            NaN         NaN   

   runtime_minutes genres  
0              NaN    NaN  
1              NaN    NaN  
2              NaN    NaN  
3              NaN    NaN  
4              NaN    NaN

*   The two datasets contain similar and related columns which make them ideal for concatenation.

*   This method works really well when you want to add more movie entries to your dataset   


#**Data Cleaning**

##Checking for duplicates

In [43]:
duplicates = combined_df[combined_df.duplicated()]
duplicates

Empty DataFrame
Columns: [title, studio, domestic_gross, foreign_gross, year, tconst, primary_title, original_title, start_year, runtime_minutes, genres]
Index: []

In [44]:
duplicates.shape

(0, 11)

As seen above, the dataframe has no duplicates

Let's try a different approach and ckeck for duplicate values in specific columns

In [45]:
combined_df[combined_df.duplicated(subset=['title', 'year'])]

title studio  domestic_gross foreign_gross  year     tconst  \
3388     NaN    NaN             NaN           NaN   NaN  tt0066787   
3389     NaN    NaN             NaN           NaN   NaN  tt0069049   
3390     NaN    NaN             NaN           NaN   NaN  tt0069204   
3391     NaN    NaN             NaN           NaN   NaN  tt0100275   
3392     NaN    NaN             NaN           NaN   NaN  tt0111414   
...      ...    ...             ...           ...   ...        ...   
149526   NaN    NaN             NaN           NaN   NaN  tt9916538   
149527   NaN    NaN             NaN           NaN   NaN  tt9916622   
149528   NaN    NaN             NaN           NaN   NaN  tt9916706   
149529   NaN    NaN             NaN           NaN   NaN  tt9916730   
149530   NaN    NaN             NaN           NaN   NaN  tt9916754   

                                      primary_title  \
3388                One Day Before the Rainy Season   
3389                     The Other Side of the Wind   
3390                                Sabse Bada Sukh   
3391                       The Wandering Soap Opera   
3392                                    A Thin Life   
...                                             ...   
149526                          Kuambil Lagi Hatiku   
149527  Rodolpho Teóphilo - O Legado de um Pioneiro   
149528                              Dankyavar Danka   
149529                                       6 Gunn   
149530               Chico Albuquerque - Revelações   

                                     original_title  start_year  \
3388                                Ashad Ka Ek Din      2019.0   
3389                     The Other Side of the Wind      2018.0   
3390                                Sabse Bada Sukh      2018.0   
3391                          La Telenovela Errante      2017.0   
3392                                    A Thin Life      2018.0   
...                                             ...         ...   
149526                          Kuambil Lagi Hatiku      2019.0   
149527  Rodolpho Teóphilo - O Legado de um Pioneiro      2015.0   
149528                              Dankyavar Danka      2013.0   
149529                                       6 Gunn      2017.0   
149530               Chico Albuquerque - Revelações      2013.0   

        runtime_minutes                genres  
3388              114.0       Biography,Drama  
3389              122.0                 Drama  
3390                NaN          Comedy,Drama  
3391               80.0  Comedy,Drama,Fantasy  
3392               75.0                Comedy  
...                 ...                   ...  
149526            123.0                 Drama  
149527              NaN           Documentary  
149528              NaN                Comedy  
149529            116.0                   NaN  
149530              NaN           Documentary  

[146143 rows x 11 columns]

In [50]:
# Remove duplicates based on title and year, keeping the first  one
combined_df = combined_df.drop_duplicates(subset=['title', 'year'], keep='first')
combined_df

title      studio  domestic_gross  \
0                                     Toy Story 3          BV     415000000.0   
1                      Alice in Wonderland (2010)          BV     334200000.0   
2     Harry Potter and the Deathly Hallows Part 1          WB     296000000.0   
3                                       Inception          WB     292600000.0   
4                             Shrek Forever After        P/DW     238700000.0   
...                                           ...         ...             ...   
3383                  Edward II (2018 re-release)          FM          4800.0   
3384                                     El Pacto        Sony          2500.0   
3385                                     The Swan  Synergetic          2400.0   
3386                            An Actor Prepares       Grav.          1700.0   
3387                                          NaN         NaN             NaN   

     foreign_gross    year     tconst primary_title original_title  \
0        652000000  2010.0        NaN           NaN            NaN   
1        691300000  2010.0        NaN           NaN            NaN   
2        664300000  2010.0        NaN           NaN            NaN   
3        535700000  2010.0        NaN           NaN            NaN   
4        513900000  2010.0        NaN           NaN            NaN   
...            ...     ...        ...           ...            ...   
3383           NaN  2018.0        NaN           NaN            NaN   
3384           NaN  2018.0        NaN           NaN            NaN   
3385           NaN  2018.0        NaN           NaN            NaN   
3386           NaN  2018.0        NaN           NaN            NaN   
3387           NaN     NaN  tt0063540     Sunghursh      Sunghursh   

      start_year  runtime_minutes              genres  
0            NaN              NaN                 NaN  
1            NaN              NaN                 NaN  
2            NaN              NaN                 NaN  
3            NaN              NaN                 NaN  
4            NaN              NaN                 NaN  
...          ...              ...                 ...  
3383         NaN              NaN                 NaN  
3384         NaN              NaN                 NaN  
3385         NaN              NaN                 NaN  
3386         NaN              NaN                 NaN  
3387      2013.0            175.0  Action,Crime,Drama  

[3388 rows x 11 columns]

Having dropped the duplicates, the number of rows have decreased while the number of columns remained the same

In [52]:
# Change all the column headings to lowercase
combined_df.columns.str.lower()


Index(['title', 'studio', 'domestic_gross', 'foreign_gross', 'year', 'tconst',
       'primary_title', 'original_title', 'start_year', 'runtime_minutes',
       'genres'],
      dtype='object')

In [53]:
# Check unique values in each column
combined_df.nunique()

title              3386
studio              257
domestic_gross     1797
foreign_gross      1204
year                  9
tconst                1
primary_title         1
original_title        1
start_year            1
runtime_minutes       1
genres                1
dtype: int64

In [54]:
# Check the missing values
df.isna()

tconst  primary_title  original_title  start_year  runtime_minutes  \
0        False          False           False       False            False   
1        False          False           False       False            False   
2        False          False           False       False            False   
3        False          False           False       False             True   
4        False          False           False       False            False   
...        ...            ...             ...         ...              ...   
146139   False          False           False       False            False   
146140   False          False           False       False             True   
146141   False          False           False       False             True   
146142   False          False           False       False            False   
146143   False          False           False       False             True   

        genres  
0        False  
1        False  
2        False  
3        False  
4        False  
...        ...  
146139   False  
146140   False  
146141   False  
146142    True  
146143   False  

[146144 rows x 6 columns]

In [60]:
# Total number of missing values
df.isna().sum()

tconst                 0
primary_title          1
original_title        22
start_year             0
runtime_minutes    31739
genres              5408
dtype: int64

In [62]:
#Check the proportion of missing values per column
missing_proportion = combined_df.isna().mean().round(3) # Calculate the missing proportion and round to 3 decimal places
missing_proportion # Display the missing proportion

title              0.000
studio             0.002
domestic_gross     0.009
foreign_gross      0.399
year               0.000
tconst             1.000
primary_title      1.000
original_title     1.000
start_year         1.000
runtime_minutes    1.000
genres             1.000
dtype: float64

In [70]:
# Get the rows with the missing values
combined_df[combined_df.isnull().any(axis=1)]

title      studio  domestic_gross  \
0                                     Toy Story 3          BV     415000000.0   
1                      Alice in Wonderland (2010)          BV     334200000.0   
2     Harry Potter and the Deathly Hallows Part 1          WB     296000000.0   
3                                       Inception          WB     292600000.0   
4                             Shrek Forever After        P/DW     238700000.0   
...                                           ...         ...             ...   
3383                  Edward II (2018 re-release)          FM          4800.0   
3384                                     El Pacto        Sony          2500.0   
3385                                     The Swan  Synergetic          2400.0   
3386                            An Actor Prepares       Grav.          1700.0   
3387                                          NaN         NaN             NaN   

     foreign_gross    year     tconst primary_title original_title  \
0        652000000  2010.0        NaN           NaN            NaN   
1        691300000  2010.0        NaN           NaN            NaN   
2        664300000  2010.0        NaN           NaN            NaN   
3        535700000  2010.0        NaN           NaN            NaN   
4        513900000  2010.0        NaN           NaN            NaN   
...            ...     ...        ...           ...            ...   
3383           NaN  2018.0        NaN           NaN            NaN   
3384           NaN  2018.0        NaN           NaN            NaN   
3385           NaN  2018.0        NaN           NaN            NaN   
3386           NaN  2018.0        NaN           NaN            NaN   
3387           NaN     NaN  tt0063540     Sunghursh      Sunghursh   

      start_year  runtime_minutes              genres  
0            NaN              NaN                 NaN  
1            NaN              NaN                 NaN  
2            NaN              NaN                 NaN  
3            NaN              NaN                 NaN  
4            NaN              NaN                 NaN  
...          ...              ...                 ...  
3383         NaN              NaN                 NaN  
3384         NaN              NaN                 NaN  
3385         NaN              NaN                 NaN  
3386         NaN              NaN                 NaN  
3387      2013.0            175.0  Action,Crime,Drama  

[3388 rows x 11 columns]

In [67]:
# Drop the rows with missing values
combined_df.dropna(axis=0, how="all", inplace=True)
combined_df

title      studio  domestic_gross  \
0                                     Toy Story 3          BV     415000000.0   
1                      Alice in Wonderland (2010)          BV     334200000.0   
2     Harry Potter and the Deathly Hallows Part 1          WB     296000000.0   
3                                       Inception          WB     292600000.0   
4                             Shrek Forever After        P/DW     238700000.0   
...                                           ...         ...             ...   
3383                  Edward II (2018 re-release)          FM          4800.0   
3384                                     El Pacto        Sony          2500.0   
3385                                     The Swan  Synergetic          2400.0   
3386                            An Actor Prepares       Grav.          1700.0   
3387                                          NaN         NaN             NaN   

     foreign_gross    year     tconst primary_title original_title  \
0        652000000  2010.0        NaN           NaN            NaN   
1        691300000  2010.0        NaN           NaN            NaN   
2        664300000  2010.0        NaN           NaN            NaN   
3        535700000  2010.0        NaN           NaN            NaN   
4        513900000  2010.0        NaN           NaN            NaN   
...            ...     ...        ...           ...            ...   
3383           NaN  2018.0        NaN           NaN            NaN   
3384           NaN  2018.0        NaN           NaN            NaN   
3385           NaN  2018.0        NaN           NaN            NaN   
3386           NaN  2018.0        NaN           NaN            NaN   
3387           NaN     NaN  tt0063540     Sunghursh      Sunghursh   

      start_year  runtime_minutes              genres  
0            NaN              NaN                 NaN  
1            NaN              NaN                 NaN  
2            NaN              NaN                 NaN  
3            NaN              NaN                 NaN  
4            NaN              NaN                 NaN  
...          ...              ...                 ...  
3383         NaN              NaN                 NaN  
3384         NaN              NaN                 NaN  
3385         NaN              NaN                 NaN  
3386         NaN              NaN                 NaN  
3387      2013.0            175.0  Action,Crime,Drama  

[3388 rows x 11 columns]